### PR TITLE
3.1 upgrade process with no chop changes

### LINKED ISSUE #3768: No-chop merge file metadata upgrade code
The changes in #3761 change the format of the file metadata stored from being just a path to a json object with a path and a range. We need to make sure we can handle the legacy format by upgrading existing entries when Accumulo is upgraded. The other option is to be backwards compatible and handle both but it would be cleaner if we could just upgrade existing entries to the new format. There's also a few other things that need upgrading as well. Below is a list that will get updated as new things are discovered.

Upgrade Tasks:

- [ ] Update format of metadata in root and metadata tablets for `DataFileColumnFamily` from a path to the new json format that contains a path and a range serialized as Base 64. The code for handling the new format of the file metadata is part of the `StoredTabletFile` class. When converting the existing file metadata (which is just a path) we should convert that to the new format which will just contain a Range that is infinite.
- [ ] Remove old [ChoppedColumnFamily](https://github.com/cshannon/accumulo/blob/ff916e6759fbfd88d09b2ff0b3681d082c1e342e/core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java#L328-L332) markers in metadata if any exist. I'm planning to move this constant over to the Upgrade section as part of https://github.com/apache/accumulo/issues/3769 .
- [ ] As noted [here](https://github.com/apache/accumulo/pull/3761#discussion_r1334818629), we also need to delete any external compaction entries which is the `ExternalCompactionColumnFamily`

### PR DESCRIPTION
**Pull request by EdColeman**
Resolves #3768.  Completes all three sub-tasks:

 - modifies StoredTabletFile references to new json encoding
 - removed obsolete references that may contain old file reference format
   - chopped references
   - external compaction references.

The obsolete references are not normally expected during an upgrade.  Upgrades will not start if there are outstanding FATE transactions - so if those references are present, it would have been from an operation that did not complete cleanly and were not cleaned up.  The will not be from an active operation.  The obsolete references are logged for information and to help determine if those operations should be re-run at some point in the future.

### CODE REVIEW

**Comment by ctubbsii** [comment]
> removal of `ExternalCompactionColumnFamily` needs to be implemented

I don't think that's related to the no-chop merge change, is it? Or is that a separate thing that was removed in 3.1?

**Comment by EdColeman** [comment]
Also verified that a 3.0 --> 3.1 upgrade with an uno instance works.

**Comment by cshannon** [comment]
@EdColeman - I have started to review this but probably won't finish today however I plan to work this weekend so should finish then. I am going to spend some time testing it myself with Uno, etc to make sure it all looks good.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (cshannon) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:139]: Right now the conversion is happening in two steps where the metadata is first scanned to load the file references into a map and then that is iterated over and updated. One thing I am wondering is if this is a problem if there's a lot of files in the system. I'm not sure what a practical upper bounds would be since depends on a lot of factors but if it's say hundreds of thousands or millions that would be a lot to load at once. We could try and break it up maybe into batches (maybe a tablet at a time or some sort of limit on files at one time and keep track of where you left off). 

I haven't really tried this before also thinking maybe we just modify the files while scanning in the same loop to avoid having to load into a map. The main thing I am not sure of is whether or not that would cause issues with the updates and we'd need to use an isolated scanner but that also buffers the entire row into memory as well. 

@keith-turner any thoughts on this?

**Review by cshannon — CHANGES_REQUESTED** [review]
@EdColeman - I tested this out with Uno and it worked well, so overall looking pretty good. I saw it clean up the chopped references and upgrade all the metadata succesfully. I added some in line comments for a few things to tweak.

The only other main thing I think this is missing is testing. There's no tests right now for any of the upgrade code in `Upgrader11to12` so I think that is important to add to make sure it works before merging this. 

Ideally we would also add an integration test or 2 as well to test the upgrade. It would be a little tricky as it would probably require having to take tables offline and manually modifying the metadata to the old format and stopping the mini accumulo cluster and then changing metadata so it would upgrade on a restart. So I'm not sure about doing an IT but the Upgrader11to12 class itself at least should have tests.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java]: Addressed in e8ae63227647

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java]: With refactoring, this method was not needed outside of the class and the method was removed in e8ae63227647

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java
Comment (EdColeman) [reply core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java]: Added contains checks in e8ae63227647

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Changed in e8ae63227647 (along with other changes)

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:197]: Added in e8ae63227647

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:96]: Changed so that the upgrade file conversion, chop and external compaction removals are done in a single pass through the metadata without creating a map of the file references in e8ae63227647. This reduced the three passes through the metadata that was previous done on first implementation to a single pass and no intermediate storage is used.

Did not use an IsolatedScanner - not sure if that is necessary or not.

Comment (cshannon) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:96]: I like these changes a lot, this will be much more efficient in terms of having to make a single pass and also with memory usage as the batch writer will flush as needed. 

The main thing left would be is if we need to use an `IsolatedScanner` or not to buffer the entire row into memory so the changes in place don't cause issues. The code already handles things if already upgraded so the main thing I'd want to avoid is processing the same columns in the row more than once and I am not sure if that could happen or not.

**Review by cshannon — COMMENTED** [review]
The latest changes look pretty good to me. I think the 2 main things left are to add a unit test for Upgrader11to12.java to verify things work (including failure cases or when metadata is already upgraded) and to get some clarification on whether or not using an IsolatedScanner for the metadata updates is necessary.

I also started thinking more about some ITs that we could write that could be done in another PR as part of a follow on issue to this. I am hoping we could write an IT that not only verifies that the file metadata was updated but also can verify consistency with scans (all the files and data can still be read).

**Comment by EdColeman** [comment]
Added tests for Upgrader11to12 with 3ff4d1655b432a

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java:56]: Can test the needsConversion function.

```suggestion
assertTrue(RootTabletMetadata.needsConversion(root212ZkData2Files));
```

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java:30]: Could add the following test also

 * Pass version 5 data and ensure an exception is thrown
 * Pass version 2 data to upgrade and ensure it behaves as expected

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:113]: I would throw an exception here instead of logging a warning because its completely unexpected.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:163]: Need to rethrow this exception.  The batch writer queues up mutations and does not send them immediately.  When addMuation throws an exception it may not have been caused by the most recently added mutation.  Could be from a mutation for changing a file.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:181]: Also need to throw an exception here.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:176]: The warn was so that that would stand out in the logs.  This will run once, and debug statements could be easier lost / ignored.  The removal of external compaction markers is because they would likely cause problems because they do not include the fence parameters.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:72]: Added in aa2c468a5b. Use IsolatedScanner and ZooKeeper root dump before / after.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java]: Changed in bceac27d1c

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java
Comment (EdColeman) [reply core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java:58]: Added tests in bceac27d1c

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java
Comment (EdColeman) [reply core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java:32]: Added test in bceac27d1c (used -1 as version so it should always be invalid)

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Added in bceac27d1c.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:167]: Added in bceac27d1c

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:187]: Added in bceac27d1c

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:182]: Its not expected that an external compaction would run across an upgrade and removing entries related to any that were running is not very interesting.  It could possibly log a lot of warns if there are a lot of ecomp entries and that would be a lot of noise that is not very useful IMO.

Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:182]: Currently, we block the upgrade from even starting if there are outstanding FATE transactions - do we need something equivalent to stop / block external compactions?

Also, during the upgrade only the root and the metadata tablets would be online and I don't think we would want concurrent operations occurring on those tables while the upgrade is executing.

My assumption was that because no FATEs are present, any flags present would be from operations that terminated and did not get a chance to clean-up.  I may be important to know which things did not stop cleanly - so it can be determined if post-upgrade clean-up operations, such as re-queuing the interrupted compaction is necessary.

**Review by cshannon — APPROVED** [review]
New updates LGTM, I think the only outstanding thing I saw was about the log level of debug vs warn but I don't have a strong preference either way so well leave that up to you and @keith-turner

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Changed to debug in 77dd24939a

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java:70]: Feel free to ignore this comment.  I would keep this class private

```suggestion
  private static class Data {
```

and add the the following method to `RootTabletMetadata` and have everything call it.

```java
    public static boolean needsUpgrade(final String json) {
      return Data.needsUpgrade(json);
    }
```

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadataTest.java:38]: If only we could move to JDK 17, then we could use [text blocks](https://docs.oracle.com/en/java/javase/15/text-blocks/index.html) for this and it would be a lot nicer.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:80]:
```suggestion
          "Could not read or write root metadata in ZooKeeper because of ZooKeeper exception", ex);
```

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:71]: Wondering if this log message should go after the ZK write.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:197]: Could use language that is more inclusive here.

```suggestion
   * Quick validation to see if value has been converted by checking the candidate can be parsed
```

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:211]: Using exceptions for expected situations may be slow, needs more research.  In the quick reading I did on the subject the expensive part of throwing an exception is creating the stack trace, which makes sense.  If this is happening for a large number of files and its slow, then it could make upgrade slow.  I think we need to a follow on issue for this, it may not be a problem but would be good to investigate.  Could either try to benchmark it or just avoid throwing and catching exceptions as part of normal processing.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:137]: It would be nice to generate a single mutation for all the tablets files.  That may be faster and would result in atomically updating the tablets files. However I don't think this has to be done.  If it were done, it could be in a follow on PR. What is here seems correct and is probably fast enough for the one time upgrade case.

**Review by keith-turner — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/RootTabletMetadata.java]: Addressed in cc6b6e25801c

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Fixed in cc6b6e25801c

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Changed in cc6b6e25801c

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Fixed in cc6b6e25801c

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:137]: I worried about memory usage for a large number of files as a single mutation - but this would also eliminate the need to check for the file needs conversion issue comment on above.

The `fileNeedsConversion` check was added to handle partial upgrades where some files had been converted and then an error would halt the upgrade leaving some files converted, some not.  The halt was trigger by errors during development, but seemed that it could allow an admin to perform corrective action and the "restart" the upgrade if it happened in general.

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java]: Modified the needs conversion check in 882be1e05f8

Changed the check to look for a starting `{` and an ending `}`.  The data written to the metadata is well controlled, so it may be enough that it looks "json like" instead of trying fully to parse.  If the column qualifier data is a plain file reference (2.1) or something that looks like a json object starting with 3.1 and the no chop fence encoding then that is likely sufficient for what this is trying to do in a upgrade.

**Review by keith-turner — APPROVED** [review]

File: server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java
Comment (EdColeman) [reply server/manager/src/main/java/org/apache/accumulo/manager/upgrade/Upgrader11to12.java:136]: For the upgrade process, there is no difference if the upgrade code were to fail on mutation updates if it was for a mutation within a tablet or for a whole tablet mutation - that is, there should be no impact to the upgrade code being re-run with a partial upgrade that had previously completed at any stage.  

Being that the upgrade code is running before things are brought online, any inconsistency in file representations needs to be resolved before the next level can be brought online.  For example, until all root table file references pointing to the metadata tablet files are resolved, the upgrade will not progress to the point of processing the metadata file references.  This is independent if the inconsistencies were within a tablet of for any tablet in general.

**Pull request merged**
